src/ffi/mod.rs

```diff
@@ -838,6 +843,83 @@ extern "C" {
     fn fortify_source();
 }
 
+#[allow(unused)]
+struct LogCensor {
+    ip_mask_seed: [u8; 32],
+    ip_regex: Regex,
+}
+
+#[allow(unused)]
+impl LogCensor {
+    fn new() -> Option<Self> {
+        RegexBuilder::new(
+            r"\b
+                (?:
+                    (?:
+                        25[0-5]
+                        |  2[0-4][0-9]
+                        |  1[0-9]{2}
+                        |  [1-9]?[0-9]
+                    )\.
+                ){3}
+                (?:
+                25[0-5]
+                |  2[0-4][0-9]
+                |  1[0-9]{2}
+                |  [1-9]?[0-9]
+                )
+                \b
+                |
+                (?:
+                    ([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}
+                |   :(:[0-9a-fA-F]{1,4}){1,7}
+                |   ([0-9a-fA-F]{1,4}:){1}(:[0-9a-fA-F]{1,4}){1,6}
+                |   ([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}
+                |   ([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}
+                |   ([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}
+                |   ([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}
+                |   ([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}
+                |   ([0-9a-fA-F]{1,4}:){1,7}:
+                )",
+        )
+        .ignore_whitespace(true)
+        .build()
+        .ok()
+        .map(|re| LogCensor {
+            ip_mask_seed: rand::thread_rng().gen::<[u8; 32]>(),
+            ip_regex: re,
+        })
+    }
+
+    fn censor_logs<'a>(&self, log: &'a str) -> std::borrow::Cow<'a, str> {
+        // Gather all of the IPs
+        self.ip_regex.replace_all(log, |captures: &Captures| {
+            // First capture is guaranteed to be present by regex crate
+            #[allow(index_access_check)]
+            let ip_match = &captures[0];
+            IpAddr::from_str(ip_match)
+                .map(|ip_addr| {
+                    // Probably good enough protection for this usecase
+                    let mut data_to_hash = vec![];
+                    match ip_addr {
+                        IpAddr::V4(ipv4) => {
+                            data_to_hash.extend_from_slice(ipv4.octets().as_slice())
+                        }
+                        IpAddr::V6(ipv6) => {
+                            data_to_hash.extend_from_slice(ipv6.octets().as_slice())
+                        }
+                    };
+                    data_to_hash.extend_from_slice(&self.ip_mask_seed);
+                    // MD5 hash is much bigger than 8 bytes
+                    #[allow(index_access_check)]
+                    let prefix = &md5::compute(data_to_hash).0[..8];
```

Comment:
Switching from sha256 to md5 is not speeding this up by much. When you compare sha256 and md5 on ip like inputs md5 is *only* 60% faster. Ahash is more than 40 **times* as fast as md5 and 70 **times* as fast as sha256:

```
ahash                   time:   [2.3799 µs 2.3920 µs 2.4061 µs]
md5                     time:   [106.22 µs 106.90 µs 107.69 µs]
sha256                  time:   [170.74 µs 171.77 µs 172.99 µs]

str_ahash               time:   [3.0648 µs 3.0818 µs 3.1020 µs]
str_md5                 time:   [108.12 µs 108.67 µs 109.25 µs]
str_sha256              time:   [175.05 µs 176.11 µs 177.31 µs]
```

See [here](https://github.com/tomaszklak/hash-perf) for details.

Reply:
@tomaszklak yeah, I did it with md5 because @packgron wanted that to be more consistent with hashing it lana, not to make it faster

Reply:
I see, in that case **if** we want to keep it as it is I think a comment explaining why md5 was chosen here would be a good idea.

On the other hand I'm not convinced that the lana precedent should be a decisive here. Since it's not using any salting, it's pretty useless (in terms of security/hiding them) for ipv4 addresses.

Reply:
Ok, so we are not using md5 - so I guess I'll make a SHA version for now. `ahash` is quick, but it feels a bit wrong to me to use for kinda security purposes hash which is meant to be used for hashmaps _exclusively_ (https://github.com/tkaitchuck/aHash?tab=readme-ov-file#ahash-------).

Reply:
I still think that using cryptographically secure hash function here is an overkill. If we have salt and truncate the output I don't see how the result could be reversed.